### ISSUE TITLE
Texture "pass through" to occluded faces when texture mapping

### ISSUE THREAD

**Issue by Woolseyyy** [cdcseacave/openMVS#903]
**Describe the bug**
It seems that the texture color pass through the front faces and map to both the front faces and the occluded faces when texture mapping.
For example:
![image](https://user-images.githubusercontent.com/19527291/201860505-4a6436b5-6b19-4d2f-8b70-385771faa320.png)

**Comment by cdcseacave** [comment]
The occlusion is normally taken into account; pls share debug data (scene and images)

**Comment by Woolseyyy** [comment]
> The occlusion is normally taken into account; pls share debug data (scene and images)

https://www.dropbox.com/s/bsd8jgm17r01dys/hotdog_my.7z?dl=0

**Comment by Woolseyyy** [comment]
I have another problem... My purpose is to map recovered diffuse albedo and roughness map to a given mesh. However, a new obj would be produced at TextureMesh. What' s worse, I find that if I run TextureMesh with the same command and different images twice, the uv of the produced mesh is different. Hence, my roughness map and diffuse albedo cannot be shared with the same mesh...

**Comment by Woolseyyy** [comment]
> The occlusion is normally taken into account; pls share debug data (scene and images)

I think I find the reason. It seems there are tinny bias of my mesh or camera poses so when I map texture, a little texture near the boundary of the mesh would pass through. At the same time, I set `--global-seam-leveling` and `--local-seam-leveling` as 1 so the middle area would be blended. Hence, a big area is "passed through".

**Comment by cdcseacave** [comment]
nor sure what you mean by pass through, I did run texturing on my machine and all looks good; how do you know it is wrong?

![alb](https://user-images.githubusercontent.com/2802670/201879718-62dcdfb6-e5d3-4329-9f9c-dbb998965d39.gif)

pls use latest develop branch, and the command line u use should simply be:
```
InterfaceCOLMAP .
TextureMesh scene.mvs --mesh-file scene_mesh.ply
```
where `scene_mesh.ply` is the known mesh

As for different uv mapping, I think that is caused by multi thrreading, pls try using only one thread `--max-threads 1`

**Comment by Woolseyyy** [comment]
> nor sure what you mean by pass through, I did run texturing on my machine and all looks good; how do you know it is wrong?
> 
> ![alb](https://user-images.githubusercontent.com/2802670/201879718-62dcdfb6-e5d3-4329-9f9c-dbb998965d39.gif) [ ![alb](https://user-images.githubusercontent.com/2802670/201879718-62dcdfb6-e5d3-4329-9f9c-dbb998965d39.gif) ](https://user-images.githubusercontent.com/2802670/201879718-62dcdfb6-e5d3-4329-9f9c-dbb998965d39.gif) [ ](https://user-images.githubusercontent.com/2802670/201879718-62dcdfb6-e5d3-4329-9f9c-dbb998965d39.gif)
> 
> pls use latest develop branch, and the command line u use should simply be:
> 
> ```
> InterfaceCOLMAP .
> TextureMesh scene.mvs --mesh-file scene_mesh.ply
> ```
> 
> where `scene_mesh.ply` is the known mesh
> 
> As for different uv mapping, I think that is caused by multi thrreading, pls try using only one thread `--max-threads 1`

Can you have a check with "roughness" folder? I meet the problem mainly with the roughness data.
Thank you about the answer about uv mappling! I will have a try.

**Comment by cdcseacave** [comment]
the algorithm is the same regardless of what the images contain

**Comment by Woolseyyy** [comment]
> --max-threads 1

ok... In my results, it still remains some "pass through" at the bottom of bread:

![image](https://user-images.githubusercontent.com/19527291/201883389-ca01e3ec-c3b2-429f-ba0d-cf689de81062.png)

**Comment by Woolseyyy** [comment]
> nor sure what you mean by pass through, I did run texturing on my machine and all looks good; how do you know it is wrong?
> 
> ![alb](https://user-images.githubusercontent.com/2802670/201879718-62dcdfb6-e5d3-4329-9f9c-dbb998965d39.gif) [ ![alb](https://user-images.githubusercontent.com/2802670/201879718-62dcdfb6-e5d3-4329-9f9c-dbb998965d39.gif) ](https://user-images.githubusercontent.com/2802670/201879718-62dcdfb6-e5d3-4329-9f9c-dbb998965d39.gif) [ ](https://user-images.githubusercontent.com/2802670/201879718-62dcdfb6-e5d3-4329-9f9c-dbb998965d39.gif)
> 
> pls use latest develop branch, and the command line u use should simply be:
> 
> ```
> InterfaceCOLMAP .
> TextureMesh scene.mvs --mesh-file scene_mesh.ply
> ```
> 
> where `scene_mesh.ply` is the known mesh
> 
> As for different uv mapping, I think that is caused by multi thrreading, pls try using only one thread `--max-threads 1`

I tried `--max-threads 1` in InterfaceCOLMAP , ReconstructMesh , and TextureMesh, but the uv issue still remains.

**Comment by cdcseacave** [comment]
again, how do u know is not seen by an image, for ex this one
![image](https://user-images.githubusercontent.com/2802670/201908792-f450fddc-4d15-4aed-9f52-b2504f2367ee.png)

**Comment by cdcseacave** [comment]
u do not need ReconstructMesh

**Comment by Woolseyyy** [comment]
> InterfaceCOLMAP

I use 
```
InterfaceCOLMAP .  --max-threads 1
TextureMesh scene.mvs --mesh-file scene_mesh.ply  --max-threads 1
```
with the latest realse version https://github.com/cdcseacave/openMVS/releases/tag/v2.0.1
but still the uv issue remains.

**Comment by Woolseyyy** [comment]
> u do not need ReconstructMesh

I try with the latest code in 'develop' branch, and use 
```
InterfaceCOLMAP .  --max-threads 1
TextureMesh scene.mvs --mesh-file scene_mesh.ply  --max-threads 1
```
but the uv issue still remains...
The data and results are https://www.dropbox.com/s/bsd8jgm17r01dys/hotdog_my.7z?dl=0

**Comment by Woolseyyy** [comment]
About the "pass through", I think there still exsit some problems. Have a look at this case:
![image](https://user-images.githubusercontent.com/19527291/202345451-f95a4394-d57f-4e28-8060-dcdd18da95d9.png)

I think the clues are clear in this input image:
![diffuse_albedo-089](https://user-images.githubusercontent.com/19527291/202345565-112859ed-68b6-4767-9645-67c802267e1e.png)

However, the recovered results look like that the apple is projected to the plane.

The data and results are in https://www.dropbox.com/s/6hgapbjtfddmfjy/fruit.7z?dl=0

**Comment by cdcseacave** [comment]
I agree, something is not right here; do you have time to look into what images are projecting there bellow the objects? you have to check `ListCameraFaces()` where a list of images seeing each triangle is made; check some of the triangles that have wrong texture by which images are seen

**Comment by cdcseacave** [comment]
I found the problem in this case: the mesh has some triangles too big; I remeshed it and the result is ok
![image](https://user-images.githubusercontent.com/2802670/202384949-1ae8ca7a-a4af-4084-962c-46ddabd86a35.png)
[scene_mesh.zip](https://github.com/cdcseacave/openMVS/files/10029005/scene_mesh.zip)

**Comment by Woolseyyy** [comment]
> I found the problem in this case: the mesh has some triangles too big; I remeshed it and the result is ok ![image](https://user-images.githubusercontent.com/2802670/202384949-1ae8ca7a-a4af-4084-962c-46ddabd86a35.png) [scene_mesh.zip](https://github.com/cdcseacave/openMVS/files/10029005/scene_mesh.zip)

I see. Thank you very much!

**Comment by cdcseacave** [comment]
Texture pass-through to occluded faces occurs when depth testing in the software rasterizer doesn't fully resolve occlusion at grazing angles. The texture pipeline has been improved on develop.
